File: orchestrator/create_runpod.ts
#!/usr/bin/env npx tsx
/**
 * Simple RunPod Container Creator
 *
 * Creates a RunPod container and outputs the SSH connection command.
 * Usage: tsx create_runpod.ts [--no-auto-terminate] [--gpu-type "NVIDIA RTX A4000"]
 */

import dotenv from "dotenv"
import path from "node:path"
import { fileURLToPath } from "node:url"

dotenv.config({
  path: path.resolve(fileURLToPath(import.meta.url), "..", "..", ".env")
})

// ============================================================================
// Configuration
// ============================================================================

const CONFIG = {
  // RunPod API
  apiKey: process.env.RUNPOD_API_KEY!,
  baseURL: "https://rest.runpod.io/v1",

  // Container settings
  repoName: "AE-Scientist",
  repoOrg: "agencyenterprise",
  repoBranch: "main",
  sshKeySecretName: "GIT_SSH_KEY_AE_SCIENTIST_B64",

  // Hardware
  defaultGpuTypes: [
    "NVIDIA RTX A4000",
    "NVIDIA RTX A4500",
    "NVIDIA RTX 3090",
    "NVIDIA RTX A5000"
  ],
  gpuCount: 1,
  containerDiskInGb: 30,
  volumeInGb: 50,

  // Docker
  imageName: "runpod/pytorch:1.0.2-cu1281-torch280-ubuntu2404",
  ports: ["22/tcp", "8888/http"],

  // Behavior
  maxRetries: 3,
  pollIntervalMs: 5000,
  maxPollAttempts: 60 // 5 minutes max wait
}

// ============================================================================
// Types
// ============================================================================

interface GpuType {
  id: string
  count: number
  displayName: string
  securePrice: number
  communityPrice: number
  oneMonthPrice: number
  threeMonthPrice: number
  sixMonthPrice: number
  oneWeekPrice: number
  communitySpotPrice: number
  secureSpotPrice: number
}

interface CpuType {
  id: string
  displayName: string
  cores: number
  threadsPerCore: number
  groupId: string
}

interface Machine {
  minPodGpuCount: number
  gpuTypeId: string
  gpuType: GpuType
  cpuCount: number
  cpuTypeId: string
  cpuType: CpuType
  location: string
  dataCenterId: string
  diskThroughputMBps: number
  maxDownloadSpeedMbps: number
  maxUploadSpeedMbps: number
  supportPublicIp: boolean
  secureCloud: boolean
  maintenanceStart?: string
  maintenanceEnd?: string
  maintenanceNote?: string
  note?: string
  costPerHr: number
  currentPricePerGpu: number
  gpuAvailable: number
  gpuDisplayName: string
}

interface NetworkVolume {
  id: string
  name: string
  size: number
  dataCenterId: string
}

interface SavingsPlan {
  costPerHr: number
  endTime: string
  gpuTypeId: string
  id: string
  podId: string
  startTime: string
}

interface PodInfo {
  id: string
  name: string
  desiredStatus: "RUNNING" | "EXITED" | "TERMINATED"
  adjustedCostPerHr: number
  aiApiId: string | null
  consumerUserId: string
  containerDiskInGb: number
  containerRegistryAuthId: string | null
  costPerHr: string
  cpuFlavorId?: string
  dockerEntrypoint?: string[]
  dockerStartCmd?: string[]
  endpointId: string | null
  env: Record<string, string>
  gpu?: GpuType
  image: string
  interruptible: boolean
  lastStartedAt: string
  lastStatusChange: string
  locked: boolean
  machine?: Machine
  machineId: string
  memoryInGb: number
  networkVolume?: NetworkVolume
  portMappings: Record<string, number>
  ports: string[]
  publicIp: string | null
  savingsPlans?: SavingsPlan[]
  slsVersion?: number
  templateId: string | null
  vcpuCount: number
  volumeEncrypted: boolean
  volumeInGb: number
  volumeMountPath: string
}

interface CreatePodResponse {
  id: string
  name: string
  imageName: string
  gpuCount: number
  [key: string]: unknown
}

// ============================================================================
// API Functions
// ============================================================================

async function makeRequest<T>(
  endpoint: string,
  options: RequestInit = {}
): Promise<T> {
  const url = `${CONFIG.baseURL}${endpoint}`
  const response = await fetch(url, {
    ...options,
    headers: {
      Authorization: `Bearer ${CONFIG.apiKey}`,
      "Content-Type": "application/json",
      ...options.headers
    }
  })

  if (!response.ok) {
    const errorText = await response.text()
    console.log({
      response,
      errorText
    })
    let errorMessage: string
    try {
      const errorJson = JSON.parse(errorText)
      errorMessage = JSON.stringify(errorJson, null, 2)
    } catch {
      errorMessage = errorText
    }
    throw new Error(`RunPod API error (${response.status}): ${errorMessage}`)
  }

  return response.json() as Promise<T>
}

async function createPod(
  gpuType: string,
  autoTerminate: boolean,
  branch: string
): Promise<CreatePodResponse> {
  const dockerStartCmd = buildDockerStartCommand(autoTerminate)

  const podPayload = {
    name: `${CONFIG.repoName}-${Date.now()}`,
    imageName: CONFIG.imageName,
    cloudType: "SECURE",
    gpuCount: CONFIG.gpuCount,
    gpuTypeIds: [gpuType],
    containerDiskInGb: CONFIG.containerDiskInGb,
    volumeInGb: CONFIG.volumeInGb,
    env: {
      GIT_SSH_KEY_B64: `{{ RUNPOD_SECRET_${CONFIG.sshKeySecretName} }}`,
      REPO_NAME: CONFIG.repoName,
      REPO_ORG: CONFIG.repoOrg,
      REPO_BRANCH: branch,
      REPO_STARTUP_CMD: ""
    },
    ports: CONFIG.ports,
    dockerStartCmd: ["bash", "-c", dockerStartCmd]
  }

  return await makeRequest<CreatePodResponse>("/pods", {
    method: "POST",
    body: JSON.stringify(podPayload)
  })
}

async function getPod(podId: string): Promise<PodInfo> {
  return await makeRequest<PodInfo>(`/pods/${podId}`)
}

async function getPodHostId(podId: string): Promise<string | null> {
  const query = `
    query pod($input: PodFilter!) {
      pod(input: $input) {
        machine {
          podHostId
        }
        __typename
      }
      __typename
    }
  `

  const response = await fetch("https://api.runpod.io/graphql", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${CONFIG.apiKey}`,
      "Content-Type": "application/json"
    },
    body: JSON.stringify({
      operationName: "pod",
      variables: {
        input: {
          podId
        }
      },
      query
    })
  })

  if (!response.ok) {
    const errorText = await response.text()
    const status = response.status
    console.warn(
      `⚠️  Failed to fetch podHostId from GraphQL API: ${status} ${errorText}`
    )
    return null
  }

  const data = await response.json()
  return data?.data?.pod?.machine?.podHostId as string | null
}

// ============================================================================
// Helper Functions
// ============================================================================

function buildDockerStartCommand(autoTerminate: boolean): string {
  const scriptParts: string[] = [
    "set -euo pipefail",
    "",
    "# === Repository Setup ===",
    "curl -fsSL https://raw.githubusercontent.com/agencyenterprise/AE-Scientist-infra/refs/heads/main/setup_repo.sh | bash",
    "",
    "# === Keep container running ===",
    'echo "Container ready! Keeping alive..."'
  ]

  if (autoTerminate) {
    scriptParts.push("")
    scriptParts.push("# Note: Container will auto-terminate on exit")
  } else {
    scriptParts.push(
      'echo "Container will stay alive until manually terminated"'
    )
    scriptParts.push("# Keep container running indefinitely")
    scriptParts.push("tail -f /dev/null")
  }

  return scriptParts.join("\n").trim()
}

async function waitForPodReady(
  podId: string
): Promise<{ pod: PodInfo; podHostId: string }> {
  console.log("\n⏳ Waiting for pod to be ready...")

  for (let attempt = 1; attempt <= CONFIG.maxPollAttempts; attempt++) {
    await new Promise((resolve) => setTimeout(resolve, CONFIG.pollIntervalMs))

    try {
      const pod = await getPod(podId)
      const isRunning = pod.desiredStatus === "RUNNING"
      const hasPublicIp = pod.publicIp !== null && pod.publicIp !== undefined
      const hasPortMappings = Object.keys(pod.portMappings || {}).length > 0

      if (isRunning && hasPublicIp && hasPortMappings) {
        const podHostId = await getPodHostId(podId)
        if (!podHostId) {
          throw new Error("Pod host ID not found after pod became ready")
        }
        console.log(
          `✅ Pod is ready! (attempt ${attempt}/${CONFIG.maxPollAttempts})`
        )
        return {
          pod,
          podHostId
        }
      }

      process.stdout.write(
        `\r   Attempt ${attempt}/${CONFIG.maxPollAttempts} booting pod...`
      )
    } catch (error) {
      console.log(`\n⚠️  Error checking pod status: ${error}`)
    }
  }

  throw new Error("Pod did not become ready in time")
}

function extractSSHInfo(
  pod: PodInfo,
  podHostId: string
): {
  command: string | null
  port: number
  ip: string
} | null {
  // Check if SSH port (22) is mapped
  const sshPublicPort = pod.portMappings?.["22"]

  if (!sshPublicPort || !pod.publicIp) {
    return null
  }

  // RunPod SSH proxy connection (using podHostId from GraphQL API)
  const command = podHostId
    ? `ssh ${podHostId}@ssh.runpod.io -i ~/.ssh/id_ed25519`
    : null


  return {
    command,
    port: sshPublicPort,
    ip: pod.publicIp
  }
}

async function createPodWithRetry(
  gpuTypes: string[],
  autoTerminate: boolean,
  branch: string
): Promise<CreatePodResponse> {
  let lastError: Error | null = null
  let attemptCount = 0
  const maxAttempts = Math.max(CONFIG.maxRetries, gpuTypes.length)

  for (let i = 0; i < maxAttempts; i++) {
    const gpuType = gpuTypes[i % gpuTypes.length]
    attemptCount++

    try {
      console.log(
        `\n[Attempt ${attemptCount}/${maxAttempts}] Creating pod with GPU: ${gpuType}`
      )
      const pod = await createPod(gpuType, autoTerminate, branch)
      console.log(`✅ Pod created successfully!`)
      console.log(`   Pod ID: ${pod.id}`)
      console.log(`   Pod name: ${pod.name}`)
      console.log(`   GPU type: ${gpuType}`)
      return pod
    } catch (error) {
      lastError = error as Error
      const errorMessage =
        error instanceof Error ? error.message : String(error)

      const isGpuUnavailableError =
        errorMessage
          .toLowerCase()
          .includes("no instances currently available") ||
        errorMessage.includes("500")

      if (isGpuUnavailableError && i < maxAttempts - 1) {
        console.log(
          `⚠️  GPU type "${gpuType}" unavailable. Trying next GPU type...`
        )
        await new Promise((resolve) => setTimeout(resolve, 1000))
        continue
      } else {
        throw error
      }
    }
  }

  throw new Error(
    `Failed to create pod after ${attemptCount} attempts. ` +
      `Tried GPU types: ${gpuTypes.join(", ")}. ` +
      `Last error: ${lastError?.message || "Unknown error"}`
  )
}

// ============================================================================
// Main Function
// ============================================================================

async function main() {
  console.log("🚀 RunPod Container Creator")
  console.log("=".repeat(50))

  // Parse command line arguments
  const args = process.argv.slice(2)
  const autoTerminate = args.includes("--auto-terminate")
  const gpuTypeArg = args.find((arg) => arg.startsWith("--gpu-types="))
  const branch = args.find((arg) => arg.startsWith("--branch="))
    ? args.find((arg) => arg.startsWith("--branch="))?.split("=")[1]
    : CONFIG.repoBranch
  const gpuTypes = gpuTypeArg
    ? gpuTypeArg.split("=")[1].split(",")
    : CONFIG.defaultGpuTypes

  console.log(`\nConfiguration:`)
  console.log(`  Repository: ${CONFIG.repoOrg}/${CONFIG.repoName}`)
  console.log(`  Branch: ${CONFIG.repoBranch}`)
  console.log(`  GPU Types: ${gpuTypes.join(", ")}`)
  console.log(`  Auto-terminate: ${autoTerminate ? "Yes" : "No"}`)

  // Validate API key
  if (!CONFIG.apiKey) {
    throw new Error(
      "RUNPOD_API_KEY environment variable is required. " +
        "Set it in your .env file or export it in your shell."
    )
  }

  // Create pod
  const pod = await createPodWithRetry(gpuTypes, autoTerminate, branch)

  // Wait for pod to be ready
  const { pod: readyPod, podHostId } = await waitForPodReady(pod.id)

  // Fetch podHostId from GraphQL API for SSH proxy connection
  console.log("\n🔍 Fetching SSH connection details...")

  // Extract SSH info
  const sshInfo = extractSSHInfo(readyPod, podHostId)

  console.log("\n" + "=".repeat(50))
  console.log("🎉 Pod is ready!")
  console.log("=".repeat(50))
  console.log(`\nPod ID: ${readyPod.id}`)
  console.log(`Pod Name: ${readyPod.name}`)
  console.log(`Public IP: ${readyPod.publicIp}`)

  if (sshInfo) {
    console.log(`\n📡 SSH Connection (RunPod Proxy - Recommended):`)
    console.log(`   ${sshInfo.command}`)
    console.log(`\n   Public IP: ${sshInfo.ip}`)
    console.log(`   SSH Port: ${sshInfo.port}`)
  } else {
    console.log("\n⚠️  SSH port not found. Pod may still be initializing.")
  }

  console.log(`\n🌐 RunPod Console:`)
  console.log(`   https://www.runpod.io/console/pods`)

  if (!autoTerminate) {
    console.log(`\n⚠️  Remember to manually terminate the pod when done!`)
  }

  console.log("\n" + "=".repeat(50))
}

// ============================================================================
// Entry Point
// ============================================================================

main()
  .then(() => {
    console.log("\n✅ Done!")
    process.exit(0)
  })
  .catch((error) => {
    console.error("\n❌ Error:", error.message)
    process.exit(1)
  })

File: server/create_runpod.py
#!/usr/bin/env python3
"""
RunPod Creator and Configurator

Creates a RunPod container, validates GPU, clones the repository,
sets up the environment, and runs the installation script.
"""

import base64
import os
import subprocess
import sys
import time
from typing import Any

import requests
from dotenv import load_dotenv


class RunPodError(Exception):
    """RunPod API error"""

    def __init__(self, message: str, status: int = 0):
        super().__init__(message)
        self.status = status


class RunPodCreator:
    def __init__(self, api_key: str):
        self.api_key = api_key
        self.base_url = "https://rest.runpod.io/v1"
        self.graphql_url = "https://api.runpod.io/graphql"

    def _make_request(
        self, endpoint: str, method: str = "GET", data: dict[str, Any] | None = None
    ) -> dict[str, Any]:
        """Make a request to the RunPod REST API"""
        url = f"{self.base_url}{endpoint}"
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

        response = requests.request(method=method, url=url, headers=headers, json=data)

        if not response.ok:
            error_text = response.text
            raise RunPodError(
                f"RunPod API error ({response.status_code}): {error_text}",
                status=response.status_code,
            )

        result: dict[str, Any] = response.json()
        return result

    def _graphql_request(self, query: str, variables: dict[str, Any]) -> dict[str, Any]:
        """Make a GraphQL request to RunPod API"""
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

        response = requests.post(
            self.graphql_url,
            headers=headers,
            json={"query": query, "variables": variables},
        )

        if not response.ok:
            raise RunPodError(
                f"GraphQL error ({response.status_code}): {response.text}",
                status=response.status_code,
            )

        result: dict[str, Any] = response.json()
        return result

    def get_pod_host_id(self, pod_id: str) -> str | None:
        """Get pod host ID for SSH proxy connection"""
        query = """
            query pod($input: PodFilter!) {
                pod(input: $input) {
                    machine {
                        podHostId
                    }
                }
            }
        """
        variables = {"input": {"podId": pod_id}}

        try:
            result = self._graphql_request(query=query, variables=variables)
            pod_host_id: str | None = (
                result.get("data", {}).get("pod", {}).get("machine", {}).get("podHostId")
            )
            return pod_host_id
        except Exception as e:
            print(f"⚠️  Warning: Failed to get podHostId: {e}")
            return None

    def _attempt_create_pod(
        self,
        name: str,
        image: str,
        gpu_type: str,
        env: dict[str, str],
        docker_cmd: str,
        attempt_number: int,
    ) -> dict[str, Any]:
        """Attempt to create a RunPod pod with a specific GPU type"""
        payload = {
            "name": name,
            "imageName": image,
            "cloudType": "SECURE",
            "gpuCount": 1,
            "gpuTypeIds": [gpu_type],
            "containerDiskInGb": 30,
            "volumeInGb": 50,
            "env": env,
            "ports": ["22/tcp"],
            "dockerStartCmd": ["bash", "-c", docker_cmd],
        }

        print(f"[Attempt {attempt_number}] Creating pod with GPU type: {gpu_type}")
        result = self._make_request(endpoint="/pods", method="POST", data=payload)
        return result

    def create_pod(
        self,
        name: str,
        image: str,
        gpu_types: list[str],
        env: dict[str, str],
        docker_cmd: str,
        max_retries: int = 3,
    ) -> dict[str, Any]:
        """Create a RunPod pod with GPU availability retry logic"""
        if not gpu_types:
            raise ValueError("At least one GPU type must be specified")

        last_error: Exception | None = None
        max_attempts = max(max_retries, len(gpu_types))

        # Round-robin through GPU types
        for i in range(max_attempts):
            gpu_type = gpu_types[i % len(gpu_types)]
            attempt_number = i + 1

            try:
                pod = self._attempt_create_pod(
                    name=name,
                    image=image,
                    gpu_type=gpu_type,
                    env=env,
                    docker_cmd=docker_cmd,
                    attempt_number=attempt_number,
                )
                print("✅ Pod created successfully!")
                print(f"   Pod ID: {pod['id']}")
                print(f"   Pod name: {pod['name']}")
                print(f"   GPU type: {gpu_type}")
                return pod
            except RunPodError as error:
                last_error = error
                is_unavailable_error = (
                    "no instances currently available" in str(error).lower() or error.status == 500
                )

                if is_unavailable_error and i < max_attempts - 1:
                    print(f"⚠️  GPU type '{gpu_type}' unavailable. Trying next GPU type...")
                    time.sleep(1)
                    continue
                elif not is_unavailable_error:
                    # Different kind of error, throw immediately
                    raise

        # If we get here, all attempts failed
        raise RunPodError(
            f"Failed to create pod after {max_attempts} attempts. "
            f"Tried GPU types: {', '.join(gpu_types)}. "
            f"Last error: {last_error}"
        )

    def get_pod(self, pod_id: str) -> dict[str, Any]:
        """Get pod information"""
        return self._make_request(endpoint=f"/pods/{pod_id}")

    def wait_for_pod_ready(
        self, pod_id: str, poll_interval: int = 5, max_attempts: int = 60
    ) -> dict[str, Any]:
        """Wait for pod to be running with SSH access"""
        print("\n⏳ Waiting for pod to be ready...")

        for attempt in range(1, max_attempts + 1):
            time.sleep(poll_interval)

            try:
                pod = self.get_pod(pod_id=pod_id)
                is_running = pod.get("desiredStatus") == "RUNNING"
                has_public_ip = pod.get("publicIp") is not None
                has_port_mappings = bool(pod.get("portMappings", {}))

                if is_running and has_public_ip and has_port_mappings:
                    print(f"✅ Pod is ready! (attempt {attempt}/{max_attempts})")
                    return pod

                print(f"\r   Attempt {attempt}/{max_attempts} - booting pod...", end="")
            except Exception as e:
                print(f"\n⚠️  Error checking pod status: {e}")

        raise RunPodError("Pod did not become ready in time")

    def execute_ssh_command(
        self, host: str, port: int, command: str, ssh_key_path: str
    ) -> tuple[int, str, str]:
        """Execute a command over SSH"""
        ssh_cmd = [
            "ssh",
            "-o",
            "StrictHostKeyChecking=no",
            "-o",
            "UserKnownHostsFile=/dev/null",
            "-i",
            ssh_key_path,
            "-p",
            str(port),
            f"root@{host}",
            command,
        ]

        result = subprocess.run(ssh_cmd, capture_output=True, text=True, timeout=300)
        return result.returncode, result.stdout, result.stderr


def main() -> None:
    print("🚀 RunPod Creator and Configurator")
    print("=" * 70)

    # Load environment variables from .env file
    load_dotenv()

    # Step 1: Read required environment variables
    print("\n📋 Reading environment variables...")
    runpod_api_key = os.environ.get("RUNPOD_API_KEY")
    if not runpod_api_key:
        print("❌ ERROR: RUNPOD_API_KEY environment variable is required")
        sys.exit(1)

    github_deploy_key = os.environ.get("GIT_DEPLOY_KEY")
    if not github_deploy_key:
        print("❌ ERROR: GIT_DEPLOY_KEY environment variable is required")
        sys.exit(1)

    # Handle both multi-line and one-liner format (with literal \n)
    github_deploy_key = github_deploy_key.replace("\\n", "\n")

    openai_api_key = os.environ.get("OPENAI_API_KEY", "")
    hf_token = os.environ.get("HF_TOKEN", "")

    print("✅ All required environment variables found")

    # Step 2: Encode deploy key as base64 for the pod
    print("\n🔑 Encoding GitHub deploy key...")
    github_deploy_key_b64 = base64.b64encode(github_deploy_key.encode()).decode()
    print("✅ Deploy key encoded")

    # Step 3: Build docker start command (following TypeScript implementation)
    script_parts = ["set -euo pipefail", ""]

    # Validate GPU
    script_parts.append("# === GPU Validation ===")
    script_parts.append('echo "Validating GPU..."')
    script_parts.append('nvidia-smi || { echo "❌ nvidia-smi failed"; exit 1; }')
    script_parts.append('echo "✅ GPU validated"')
    script_parts.append("")

    # Repository setup using external script (matches TypeScript approach)
    script_parts.append("# === Repository Setup ===")
    script_parts.append(
        "curl -fsSL https://raw.githubusercontent.com/agencyenterprise/AE-Scientist-infra/refs/heads/main/setup_repo.sh | bash"
    )
    script_parts.append("")

    # Create .env file with API keys
    script_parts.append("# === Environment Setup ===")
    script_parts.append('echo "Creating .env file..."')
    script_parts.append("cd /workspace/AE-Scientist/research_pipeline")
    script_parts.append("cat > .env << 'EOF'")
    script_parts.append(f"OPENAI_API_KEY={openai_api_key}")
    script_parts.append(f"HF_TOKEN={hf_token}")
    script_parts.append("EOF")
    script_parts.append("")

    # Run installation script
    script_parts.append("# === Installation ===")
    script_parts.append('echo "Running installation script..."')
    script_parts.append("cd /workspace/AE-Scientist")
    script_parts.append("bash install_run_pod.sh")
    script_parts.append("")

    script_parts.append('echo "=== Setup complete ==="')
    script_parts.append("")

    # Run research pipeline
    script_parts.append("# === Starting Research Pipeline ===")
    script_parts.append('echo "Launching research pipeline..."')
    script_parts.append("cd /workspace/AE-Scientist/research_pipeline")
    script_parts.append("source .venv/bin/activate")
    script_parts.append(
        "python launch_scientist_bfts.py bfts_config_gpt-5.yaml 2>&1 | tee -a /workspace/research_pipeline.log"
    )
    script_parts.append(
        'echo "Research pipeline completed. Check /workspace/research_pipeline.log for full output."'
    )

    docker_cmd = "\n".join(script_parts).strip()

    # Step 4: Create RunPod
    creator = RunPodCreator(api_key=runpod_api_key)

    gpu_types = [
        "NVIDIA GeForce RTX 5090",
        "NVIDIA GeForce RTX 3090",
        "NVIDIA RTX A4000",
        "NVIDIA RTX A4500",
        "NVIDIA RTX A5000",
    ]

    pod_name = f"ae-scientist-{int(time.time())}"

    # Build environment variables for the pod (matches TypeScript implementation)
    pod_env = {
        # GitHub deploy key (base64 encoded for the external setup script)
        "GIT_SSH_KEY_B64": github_deploy_key_b64,
        # Repository configuration (used by setup_repo.sh)
        "REPO_NAME": "AE-Scientist",
        "REPO_ORG": "agencyenterprise",
        "REPO_BRANCH": "main",
        "REPO_STARTUP_CMD": "",
        # API keys (will be written to .env file by our script)
        "OPENAI_API_KEY": openai_api_key,
        "HF_TOKEN": hf_token,
    }

    pod = creator.create_pod(
        name=pod_name,
        image="runpod/pytorch:1.0.2-cu1281-torch280-ubuntu2404",
        gpu_types=gpu_types,
        env=pod_env,
        docker_cmd=docker_cmd,
    )

    pod_id = pod["id"]

    # Step 5: Wait for pod to be ready
    ready_pod = creator.wait_for_pod_ready(pod_id=pod_id)

    # Step 6: Get SSH connection details
    print("\n🔍 Fetching SSH connection details...")
    public_ip = ready_pod.get("publicIp")
    ssh_port = ready_pod.get("portMappings", {}).get("22")
    pod_host_id = creator.get_pod_host_id(pod_id=pod_id)

    # Step 7: Output results
    print("\n" + "=" * 70)
    print("🎉 Pod is ready!")
    print("=" * 70)
    print(f"\nPod ID: {pod_id}")
    print(f"Pod Name: {ready_pod.get('name')}")
    print(f"Public IP: {public_ip}")

    if ssh_port and public_ip:
        print("\n📡 SSH Connection:")
        if pod_host_id:
            print("  Via RunPod Proxy (Recommended):")
            print(f"    ssh {pod_host_id}@ssh.runpod.io -i ~/.ssh/id_ed25519")
        print("\n  Via Public IP:")
        print(f"    ssh root@{public_ip} -p {ssh_port} -i ~/.ssh/id_ed25519")
    else:
        print("\n⚠️  SSH port not available yet")

    print("\n🌐 RunPod Console:")
    print("   https://www.runpod.io/console/pods")

    print("\n⚠️  Remember to manually terminate the pod when done!")
    print("\n" + "=" * 70)


if __name__ == "__main__":
    main()
